### PR TITLE
Cp scope 1 architecture

### PR DESCRIPTION
**Pull request by luke-moth**
scope 1 (refactoring critical path)

### CODE REVIEW

File: src/openclsim/critical_path/dependencies_from_model.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        get dependencies from simulation setup and
+        cross reference with recorded activities to get dependencies of relevant instances
+        requires self.object_list, self.activity_list and self.recorded_activities_df
+
+        Notes
+        ---------
+        this method is likely to provide too few dependencies and might raise warnings
+        for unsupported start event conditions, activities et cetera
+
+        this method contains approx 1000 (!) lines in original commit
```
Comment (Pietervanhalem) [inline src/openclsim/critical_path/dependencies_from_model.py:29]: First word of the scentence starts with a capital.

no need to descript the number of lines of code in the docstring

File: src/openclsim/critical_path/dependencies_from_model.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        This function is preliminary, and hence has its limitations. Mostly due
+        to critical information for dependencies in the way OpenCLSim now logs
+        its information. For proper dependency checks for extracting the
+        critical path, the logging module of OpenCLSim must be extended.
+
+    """
+    return []
+
+
+def get_resource_capacity_dependencies(recorded_activities_df, objects_list):
```
Comment (Pietervanhalem) [inline src/openclsim/critical_path/dependencies_from_model.py:94]: We are making an object oriented module, no seperate functions! 

all the parameters of the functions are atributes of the the object. When you make the methods of the objects you dont have to pass the arguments. There is no reason the make these functions outside the object.

File: src/openclsim/critical_path/dependencies_from_model.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    Returns
+    --------
+    dependencies_list : list
+        list of tuples of dependencies as occured in simulation
+        (tuple of cp_activity_id) wrt resource capacity
+    """
+    return []
+
+
+# Get dependencies based on start events
```
Comment (Pietervanhalem) [inline src/openclsim/critical_path/dependencies_from_model.py:134]: why the comment

**Review by Pietervanhalem — CHANGES_REQUESTED** [review]

File: src/openclsim/critical_path/base_cp.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        DependencyGraph is wat robuuster/zou mogelijk in andere toepassingen ook worden gebruikt.
+    UNKNOWN lines for 'dependencies_from_simpy_step' (maar zeg 100-200) (SCOPE 6)
+    UNKNOWN lines for netjes simpy logging patch maken/implementeren (maar zeg 200-300) (SCOPE 6)
+- NAAMGEVINGEN
+    - simulation_graph/SimulationGraph maakt een graaf van de simulatie (dus
+    alle recorded activiteiten 'instances', afhankelijkheden en hun simulatietijden)
+    en kan het kritieke pad vinden
+    - base_cp/BaseCP: de baseclass voor het vinden van dependencies,
+    maken van overzicht recorded_activities en vinden kritieke pad.
+    - de drie methoden om dependencies te vinden hebben dan ook nog eigen module/class
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/base_cp.py:25]: ok

File: src/openclsim/critical_path/base_cp.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+"""
+OUTLINE SCOPES
+
+scope1: architectuur
+scope2: `get_recorded_activity_df`
+scope3: `get_activity_graph` + `compute_critical_path`
+scope4: `DependenciesFromLog`
+scope5: `DependenciesFromModel`
+scope6: `DependenciesFromSimpy`
+"""
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/base_cp.py:140]: ok

File: src/openclsim/critical_path/base_cp.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+    def _make_recorded_activities_df(self):
+        """
+        set a recorded_activity_df in self
+        uses the logs of provided activities and sim objects, combines these, adds unique UUID
+        and reshape into format such that single row has a start time and an end time
+
+        Notes
+        ------
+        in original commit approx 250 lines of code (split over several functions)
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/base_cp.py:78]: ok

File: src/openclsim/critical_path/base_cp.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    def __make_simulation_graph(self):
+        """use self.recorded_activity_df and self.dependency_list to build graph of
+        (interconnected) activities as evaluated in time in simulation"""
+        self.simulation_graph = SimulationGraph(
+            self.recorded_activities_df, self.dependency_list
+        )
+
+    def get_critical_path_df(self):
+        """
+        enrich recorded activity df with column 'is_critical' and return this dataframe
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/base_cp.py]: ok

File: src/openclsim/critical_path/base_cp.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        self.simulation_graph = None
+
+    @abstractmethod
+    def get_dependency_list(self):
+        """must be implemented by child classes"""
+        return []
+
+    def _make_recorded_activities_df(self):
+        """
+        set a recorded_activity_df in self
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/base_cp.py]: ok

File: src/openclsim/critical_path/simulation_graph.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    """
+
+    def __init__(self, recorded_activity_df, dependency_list):
+        """init"""
+        # set in self
+        self.recorded_activity_df = self.__check_rec(recorded_activity_df)
+        self.dependency_list = self.__check_dep(dependency_list)
+
+        # init
+        self.G = None
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/simulation_graph.py:46]: ok

File: src/openclsim/critical_path/dependencies_from_model.py (hunk 1 above)
Comment (luke-moth) [reply src/openclsim/critical_path/dependencies_from_model.py:29]: ok

File: src/openclsim/critical_path/dependencies_from_model.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+        super().__init__(*args, **kwargs)
+
+    def get_dependency_list(self):
+        """
+        get dependencies from simulation setup and
+        cross reference with recorded activities to get dependencies of relevant instances
+        requires self.object_list, self.activity_list and self.recorded_activities_df
+
+        Notes
+        ---------
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/dependencies_from_model.py:25]: ok

File: src/openclsim/critical_path/dependencies_from_model.py (hunk 3 above)
Comment (luke-moth) [reply src/openclsim/critical_path/dependencies_from_model.py:134]: I can remove ;)

File: src/openclsim/critical_path/base_cp.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+"""
+
+from abc import ABC, abstractmethod
+
+from openclsim.critical_path.simulation_graph import SimulationGraph
+
+
+class BaseCP(ABC):
+    """
+    base class for critical path
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/base_cp.py]: ok

File: src/openclsim/critical_path/base_cp.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
+"""
+module with BaseCp class that has (non abstract) methods
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/base_cp.py]: ok

File: src/openclsim/critical_path/base_cp.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+        self,
+        env,
+        object_list,
+        activity_list,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+
+        # some asserts
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/base_cp.py:36]: intended as a todo (added todo)

File: src/openclsim/critical_path/base_cp.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
+"""
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/base_cp.py:1]: yep

File: src/openclsim/critical_path/base_cp.py [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+            self.recorded_activities_df, self.dependency_list
+        )
+
+    def get_critical_path_df(self):
+        """
+        enrich recorded activity df with column 'is_critical' and return this dataframe
+        """
+        self._make_recorded_activities_df()  # makes self.recorded_activities_df
+        self.dependency_list = self.get_dependency_list()
+        self.__make_simulation_graph()  # makes self.activity_graph
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/base_cp.py:86]: ok

File: src/openclsim/critical_path/dependencies_from_model.py [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+           ``C``, and the edges ``(A, B)`` and ``(B, C)``. By replacing ``S``,
+           the resulting graph will have nodes ``A``, ``B`` and ``C``, edges
+           ``(A, B)`` and ``(B, C)``, and also the edge ``(C, A)`` providing
+           the recurrency from the while loop.
+        4. Now all activities are base activities, hence this is the final
+           form of the graph.
+
+        """
+
+    def getListDependencies(self):
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/dependencies_from_model.py:237]: true

File: src/openclsim/critical_path/dependencies_from_model.py [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+        """Return the list of dependencies based on the graph."""
+        return list(self.G.edges)
+
+    def getListBaseActivities(self):
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/dependencies_from_model.py:241]: true

File: src/openclsim/critical_path/base_cp.py [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+class BaseCP(ABC):
+    """
+    base class for critical path
+
+    Parameters
+    ------------
+    env : instance of simpy.env or instance of class that inherits from simpy.env
+    object_list : list of all (simulation) objects with Log mixin (after simulation)
+    activity_list : list of all (simulation) activities with Log mixin (after simulation)
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/base_cp.py:20]: true

File: src/openclsim/critical_path/dependencies_from_model.py [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+    def __init__(self, main_activities):
+        """Init."""
+        # initiate
+        self._main_activities = main_activities
+        self.G = nx.DiGraph()
+
+        # construct the graph
+        self._constructGraph()
+
+    def _constructGraph(self):
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/dependencies_from_model.py:211]: yep

File: src/openclsim/critical_path/simulation_graph.py [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+        contains unique identifyers for simulated activities in time. This list
+        is to contain tuples of those IDs defining dependencies between the
+        corresponding activities. E.g. if the completion of activity with
+        ``id==1`` triggers the start of activity with ``id==2``, then this
+        should appear in the list as tuple ``(1, 2)``.
+    """
+
+    def __init__(self, recorded_activity_df, dependency_list):
+        """init"""
+        # set in self
```
[the remark this answers is not in the document]
Comment (luke-moth) [reply src/openclsim/critical_path/simulation_graph.py:41]: removed

File: src/openclsim/critical_path/dependencies_from_model.py (hunk 2 above)
Comment (luke-moth) [reply src/openclsim/critical_path/dependencies_from_model.py:94]: ok - moved to class

Comment (luke-moth) [reply src/openclsim/critical_path/dependencies_from_model.py:94]: will be big one, but maybe for later discussion...

**Review by Pietervanhalem — APPROVED** [review]

**Comment by luke-moth** [comment]
@SiggyF @markvankoningsveld ik mag niet mergen. Zou een van jullie dit kunnen doen?

**Comment by luke-moth** [comment]
Of mij (tijdelijk) de rechten geven is natuurlijk ook goed

**Comment by SiggyF** [comment]
I already merged it. Let's check later how to solve this.

**Pull request merged**